src/imitation/rewards/discrim_net.py

```diff
@@ -424,11 +431,14 @@ def build_graph(self):
             self._disc_logits_gen_is_high, self._disc_mlp = self._build_discrim_net(
                 [self.obs_input, self.act_input], **self._build_discrim_net_kwargs
             )
-        self._policy_test_reward = self._policy_train_reward = -tf.log_sigmoid(
-            self._disc_logits_gen_is_high
-        )
 
         self._disc_loss = tf.nn.sigmoid_cross_entropy_with_logits(
             logits=self._disc_logits_gen_is_high,
             labels=tf.cast(self.labels_gen_is_one_ph, tf.float32),
         )
+
+        reward = tf.log_sigmoid(self._disc_logits_gen_is_high)
+        if self._positive_rewards:
+            reward *= -1
```

Comment:
I don't understand how this works. AFAIK we don't backpropagate through `reward` -- so won't flipping the sign not change the discriminator training, and so cause the policy to try to maximize dissimilarity from expert?

Reply:
Oh, I def coded their hack up wrong. Yeah, this is a pretty obvious mistake. But FWIW, we were stuck on GAIL mountain car flatlining at -200 performance for a month, and after adding in the hack, I got 1/5 seeds to reach expert and another seed to not be flatline.

Make sense given that even with an incorrectly coded version of the hack we now have negative living reward.

Reply:
What we actually want is this:
![image](https://user-images.githubusercontent.com/1750835/85470851-74411200-b564-11ea-9d05-41ab550d9bd2.png)

Reply:
OK, I'm super confused now, because looks like our default (before this PR) is equivalent to `self.favor_zero_expert_rewards is True`, which is what openai/imitation claims is better for MountainCar because it has negative rewards?

Yeah, our log sigmoid is negative rewards only.

Reply:
I haven't thought about this enough to understand what's going on, but it looks like before we had the reward being `-tf.log_sigmoid(self._disc_logits_gen_is_high)`. Now we have it being `tf.log_sigmoid(self._disc_logits_gen_is_high)` and then you flip the sign if `self._positive_rewards` is set.

So the default before the PR I don't think maps onto either of the GAIL options.

Is it possible we had a sign error this whole time? It'd explain why things work better when you flip the sign! :thinking:

Reply:
Ok, now I'm doubly super confused....

How on earth is GAIL doing even passably well if we have the wrong sign. Will be taking a look into this in the paper, the paper's code, and our experiments

Reply:
Haven't checked their code throughly yet, but I believe that their logits are "fake is high" whereas our logits are "low is high". Which would mean that before this PR, we were the equivalent of
`rewards_B = -thutil.logsigmoid(-scores_B)`.

Reply:
`logsigmoid(-x) ?= log(1 - sigmoid(x))`? In which case we were actually following the else-clause correctly?

I need to brush up on my math...

Reply:
I want to reverse the meaning of our logits so that the logits are consistent with the paper and other codebases

Reply:
Being consistent sounds good.